hooks/shared/useGetRequest/useGetRequest.tsx

```diff
@@ -0,0 +1,179 @@
+
+  // Used to prevent state update if the component is unmounted
+  const cancelRequest = useRef<boolean>(false);
+
+  const initialState: State<T> = {
+    error: undefined,
+    data: undefined,
+  };
+
+  // Keep state logic separated
+  const fetchReducer = (state: State<T>, action: Action<T>): State<T> => {
+    switch (action.type) {
+      case "loading":
+        return { ...initialState };
+      case "fetched":
+        return { ...initialState, data: action.payload, error: undefined };
+      case "error":
+        return { ...initialState, error: action.payload };
+      default:
+        return state;
+    }
+  };
+
+  const [state, dispatch] = useReducer(fetchReducer, initialState);
+  const {
+    accessToken,
+    refreshToken,
+    dispatch: authDispatch,
+  } = useAuthContext();
+
+  const fetchData = useCallback(async () => {
+    dispatch({ type: "loading" });
+
+    // If a cache exists for this url, return it
+    // If token is required, cache is automatically disabled
+    if (!options?.requiresToken && options?.useCache && cache.current[url!]) {
+      dispatch({ type: "fetched", payload: cache.current[url!] });
+      return;
+    }
+
+    if (!options?.requiresToken) {
+      if (cancelRequest.current) return;
+      const response = await fetch(url!);
+      const json = await response.json();
+
+      if (!response.ok) {
+        if (cancelRequest.current) return;
+        const error: FetchError = new Error(response.statusText);
+        error.status = response.status;
+        error.message = json?.message;
+        dispatch({ type: "error", payload: error as FetchError });
+        return;
+      }
+
+      dispatch({ type: "fetched", payload: json });
+
+      return;
+    }
+
+    try {
+      const response = await fetch(url!, {
+        headers: {
+          Authorization: `Bearer ${accessToken}`,
+        },
+      });
+      if (!response.ok) {
+        const error: FetchError = new Error(response.statusText);
+        const json = await response.json();
+        error.status = response.status;
+        error.message = json?.message;
+        throw error;
+      }
+
+      const data = (await response.json()) as T;
+      cache.current[url!] = data;
+      if (cancelRequest.current) return;
+
+      dispatch({ type: "fetched", payload: data });
+    } catch (error) {
+      if (cancelRequest.current) return;
+
+      if ((error as FetchError)?.status === 401) {
+        try {
+          const REQUEST_ACCESS_TOKEN_URL = `${process.env.NEXT_PUBLIC_BACKEND_URL}/users/api/token/refresh/`;
+
+          const requestNewAccessToken = await fetch(REQUEST_ACCESS_TOKEN_URL, {
+            method: "POST",
+            headers: {
+              "Content-Type": "application/json",
+            },
+            body: JSON.stringify({
+              refresh: refreshToken,
+            }),
+          });
+
+          const { access, refresh } = await requestNewAccessToken.json();
+          console.log(access);
```

Comment:
Hello Alex I believe you forgot to remove this :D. Please check all your files for console logs before you push them :P

Reply:
Oh yeah thanks rashad for pointing that out. I'll remove them on the next commit.